flocker/control/_model.py

```diff
@@ -451,7 +482,10 @@ def __new__(cls, hostname=None, **kwargs):
         return PClass.__new__(cls, **kwargs)
 
     uuid = field(type=UUID, mandatory=True)
-    applications = pset_field(Application)
+    applications = pmap_field(
+        unicode, Application, invariant=_keys_match("name"),
+        factory=lambda x: _turn_iterable_to_mapping_from_attribute('name', x)
```

Comment:
Motivation of turning into a pmap rather than pset: Enforces the `unique primary key`-ness of the field in the data model. Also conceptually I want to be able to do things like "add a dataset to the node with UUID foo" rather than "remove the node with UUID foo from this set, and insert a new object into the set with an additional dataset".  Encoding it this way makes it _much_ easier to write a general-purpose differ.

Motivation for the factory: Now we have duplicate data. You need to specify the name of the application as application.name _and_ as the key to the object in the mapping. Rather, when you are initially constructing a bunch of applications, just specify application.name, and this factory will turn it into a mapping. Conveniently this also saved me some trouble migrating all of the tests to the new format.

Added a constraint to Application that the type of name should be a unicode. Really it was totally unconstrained in the past.